### PR TITLE
Workaround tsuru's issue when removing multiple units

### PR DESCRIPTION
**Pull request by yagossc**
# Description

There's an issue when performing the swap of an app with multiple units. After the first unit is removed, the following error is encountered when trying to remove the remaining processes:

`event locked: app(user-tests-green) running "app.update.unit.remove" start by user yago.user@u.user at 2020-08-17T18:19:44Z`

According to tsuru's team, this is probably due to the project's lock scheme.

More about the changes can be found in the README.

# Changes Summary

- Add a number of retries to the `remove_units_per_process_type` method;
- Make it configurable (and optional) through the `.ini` file;

# How to test

Unit tests were added, so:
```
$ make testdeps  && make test
```

or, if you wish to use Docker:
```
$ make docker-test
```

One can also try swapping a test-app to see the timeouts and retries (don't forget to customize your `.ini` file as mentioned in this project's README).

### CODE REVIEW

File: src/bluegreen.py [hunk 1]
```diff
@@ -145,11 +147,28 @@ def remove_units_per_process_type(self, app, units_to_remove, process_name):
     conn = create_connection(self.target)
     conn.request("DELETE", "/apps/" + app + '/units?units=' + str(units_to_remove) + '&process=' + process_name, '', headers)
     response = conn.getresponse()
+
     if response.status != 200:
-      print "Error removing '%s' units from %s. You'll need to remove manually." % (process_name, app)
-      return False
+      for i in range(self.retry_times):
```
Comment (mauricioabreu) [inline src/bluegreen.py:152]: Here you can change the range to be `range(1, 11)`. From 1 to 10 (11 is not included)
This way you avoid the `+1` inside the loop. Makes sense?

File: src/bluegreen.py [hunk 2]
```diff
@@ -311,9 +330,9 @@ def deploy_swap(self, apps, cname):
 
     if not self.add_units(apps[1], self.total_units(apps[0])):
       return 2
-    
+
     if not self.swap(apps[0], apps[1], False):
-      print "\n  Error swaping {} and {}. Aborting...".format(apps[0], apps[1])          
+      print "\n  Error swaping {} and {}. Aborting...".format(apps[0], apps[1])
```
Comment (mauricioabreu) [inline src/bluegreen.py:335]: Good!

File: src/bluegreen.py [hunk 3]
```diff
@@ -348,6 +367,14 @@ def load(self, filepath):
       deploy_dir = config.get('Application', 'deploy_dir')
     except ConfigParser.NoOptionError:
       deploy_dir = None
+    try:
+      retry_times = int(config.get('UnitsRemoval', 'retry_times'))
+    except (ConfigParser.NoSectionError ,ConfigParser.NoOptionError, ValueError):
```
Comment (mauricioabreu) [inline src/bluegreen.py:372]: This comma seems wrong.

File: src/bluegreen.py [hunk 4]
```diff
@@ -348,6 +367,14 @@ def load(self, filepath):
...
+      retry_times = 0
```
Comment (mauricioabreu) [inline src/bluegreen.py:373]: Put a newline here to separate these two `try`.

File: src/bluegreen.py [hunk 5]
```diff
@@ -348,6 +367,14 @@ def load(self, filepath):
...
+    try:
+      retry_sleep = int(config.get('UnitsRemoval', 'retry_sleep'))
```
Comment (mauricioabreu) [inline src/bluegreen.py:375]: Aqui dá pra usar config.getint

File: src/bluegreen.py (hunk 1 above)
Comment (yagossc) [reply src/bluegreen.py:152]: Makes sense.

File: src/bluegreen.py (hunk 3 above)
Comment (yagossc) [reply src/bluegreen.py:372]: Indeed it is.

File: src/bluegreen.py (hunk 5 above)
Comment (yagossc) [reply src/bluegreen.py:375]: Thanks, this is better than the type cast.

**Comment by emerleite** [comment]
Can anybody else take a look and merge?

**Comment by yagossc** [comment]
@emerleite , we'd like to put a hold on this, because we think we might have found a bug in Tsuru.

As described, when trying to remove units consecutively we get the error:

`event locked: app(user-tests-green) running "app.update.unit.remove" start by user yago.user@u.user at 2020-08-17T18:19:44Z`

However, while this is happening, if you inquire Tsuru about the "lock status":
``` sh
curl -s https://tsuru.server.com/apps/user-tests-green \
       -H "Authorization: bearer $(cat ~/.tsuru/token)" | jq .lock
```
The reply is always false, which shouldn't be the case. We're currently waiting to hear from Tsuru's team.

**Comment by yagossc** [comment]
According to Tsuru's team, the `lock status` should be inferred by checking if there is a `running event` for the target app, and, even though, information about the lock itself is not very consistent, a `200 OK` status code as response should suffice for our needs.

The mentioned check has been added and this **PR is ready for further review and, if allowed, merge**.

**Pull request merged**